### PR TITLE
[CL Simulator] Fix existing simulator and test simulator flow for CollectFees, CollectIncentives

### PR DESCRIPTION
**Pull request by stackman27**
[Closes: #4753](https://github.com/osmosis-labs/osmosis/issues/4753)

## What is the purpose of the change

Simulator: Testing flow createPool -> createPositions -> swap Randomly -> CollectFees
- Create Pool with random values (could be multiple pools)
- createRandom Position in the above pool (could be multiple positions)
- SwapRandom amount in the above pool & position (could be across multiple pools and positions
- collectFees for random positions created above and wherever swap happened

Simulator: Testing flow createPool -> createPosition -> collectIncentives
- Create Pool with random values (could be multiple pools)
- createRandom Position in the above pool (could be multiple positions)
- Let random x amount of time pass (in simulator this could be done through the amount of blocks created)
- Collect Incentives for that position (should be for created positions)

Extra features
- fixed CreateIncentive wiring 
- added CollectIncentive simulator check
- added CreateIncentive simulator check

## Brief Changelog
- registered CreateIncentive message and added CL module simulator test

## Testing and Verifying
added sim test

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: proto/osmosis/concentrated-liquidity/tx.proto
Comment (stackman27) [inline proto/osmosis/concentrated-liquidity/tx.proto:18]: note to reviewers: idk why this was missed while creating the msg

File: x/pool-incentives/keeper/keeper.go
Comment (stackman27) [inline x/pool-incentives/keeper/keeper.go:115]: note to reviewers: we technically donot care about the denom since, we donot use it while creating CL gauges. so i changed it to "stake" to make simulator testing easier. We cannot leave it empty as it is a require field in `CreateGauge`

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/simulation/sim_msgs.go:97]: This is how I see this working best:

1. Randomly select a CL pool
2. Return all CL position IDs that are currently in that pool
3. Randomly select one of these CL positions
4. Withdraw a random amount from this position

Creating a position every time we withdraw feels like we could be missing out on some important edge cases

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/simulation/sim_msgs.go:219]: Same here

File: x/pool-incentives/keeper/keeper.go
Comment (czarcas7ic) [inline x/pool-incentives/keeper/keeper.go:112]:
```suggestion
		// lockQueryType as byTime. Although we do not need this check, we still cannot pass empty struct.
```

**Review by czarcas7ic — CHANGES_REQUESTED** [review]

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (stackman27) [reply x/concentrated-liquidity/simulation/sim_msgs.go:97]: I see, i originally thought we wanted to test the "entire flow", for instance; 
createPool -> createPosition -> performSwaps -> collect fees 
while randomizing the createPool parameters, selecting a random position from that position, performing lots of swaps and collecting random fees for random positions

Comment (stackman27) [reply x/concentrated-liquidity/simulation/sim_msgs.go:97]: but understandably your methods will create more randomness, which also makes sense

File: x/concentrated-liquidity/lp.go
Comment (stackman27) [reply x/concentrated-liquidity/lp.go]: fixed

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (stackman27) [reply x/concentrated-liquidity/simulation/sim_msgs.go]: changed

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (stackman27) [reply x/concentrated-liquidity/simulation/sim_msgs.go]: changed

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/simulation/sim_msgs.go:83]: The flow should be the following:

1. Get a random CL pool
2. Utilize the PoolId to PositionId mapping. This will allow you to iterate that store and select a random position ID form that list
3. Use this position to attempt to withdraw

What you implemented does the following:
1. Gets a random address
2. Hopes the address has a position

^ this will result in lots of failed attempts to withdraw, while the first option will fail much less

Comment (stackman27) [reply x/concentrated-liquidity/simulation/sim_msgs.go]: i see, the poolId -> positionid mapping flow was def very clean! 

changed!

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/simulation/sim_msgs.go:99]: Is this no longer a todo?

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/simulation/sim_msgs.go:75]:
```suggestion
		return nil, fmt.Errorf("user does not have any positions")
```

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/simulation/sim_msgs.go:90]: This check isn't necessary right? We are pulling the liquidity directly from the position struct and using that as the max for the withdrawAmount

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/simulation/sim_msgs.go:137]: Why don't we do the same thing as above with checking positionIds in a pool ID and go from there?

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/simulation/sim_msgs.go:142]:
```suggestion
		return nil, fmt.Errorf("user does not have pool tokens")
```

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/simulation/sim_msgs.go:103]: Can we remove fullFlow from the name here

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/simulation/sim_msgs.go:376]:
```suggestion
	// Retrieve minTick and maxTick from precision factor
```

**Review by czarcas7ic — CHANGES_REQUESTED** [review]

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (stackman27) [reply x/concentrated-liquidity/simulation/sim_msgs.go:90]: yea this was just an extra check i added, i'll rerun the sim to see if this has any impact. And if not i'll remove it

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (stackman27) [reply x/concentrated-liquidity/simulation/sim_msgs.go:137]: because for MsgFeesRequest we need senders position Ids so, i thought it would make more sense to get random position for that specific sender instead of random position in a pool

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/simulation/sim_msgs.go:154]: I am not quite understanding the logic here. Why are we swapping until one of the tokens runs out?

**Comment by czarcas7ic** [comment]
Just pinging here that one of the gotests are failing it seems like.

Also, are you running the simulator and writing results to the .db file and exploring the results? I think that would be the best way to determine if we are getting good throughput (or if one message is almost always failing, etc)

Lmk if you need help doing the above, would be happy to show you!

**Comment by stackman27** [comment]
> Just pinging here that one of the gotests are failing it seems like.
> 
> Also, are you running the simulator and writing results to the .db file and exploring the results? I think that would be the best way to determine if we are getting good throughput (or if one message is almost always failing, etc)
> 
> Lmk if you need help doing the above, would be happy to show you!


fixed the gotest! 

yup i exported the results and this is the output. The collectFees mostly fails with this error;
```
{"entry_kind":"msg","height":4,"order":31,"operation":{"route":"concentratedliquidity","name":"CollectFees","comment":"unable to build msg due to: invalid sqrt price limit given (0.000000001000000000), should be greater than (0.000000001000000000) and less than (0.000000000000000000)","ok":false,"gas_wanted":0,"gas_used":0,"msg":null}}
```

```
"concentratedliquidity": {
  "CollectFees": {
   "failure": 95,
   "ok": 11
  },
  "CollectIncentives": {
   "failure": 62,
   "ok": 61
  },
  "CreateConcentratedPool": {
   "failure": 2,
   "ok": 110
  },
  "CreateIncentives": {
   "failure": 16,
   "ok": 96
  },
  "CreatePosition": {
   "failure": 11,
   "ok": 91
  },
  "WithdrawPosition": {
   "failure": 67,
   "ok": 37
  }
 },
```

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (stackman27) [reply x/concentrated-liquidity/simulation/sim_msgs.go]: this is basically swapping until the owner runs out of funds, i removed token1Amt calculations to make things clearer. So now, user swaps tokenA -> tokenB until their tokenA funds are <= 0

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/simulation/sim_msgs.go:350]: user c?

**Review by czarcas7ic — APPROVED** [review]
Nice work! :)

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (stackman27) [reply x/concentrated-liquidity/simulation/sim_msgs.go:330]: oh wow, i didn't know that. So if a user wants to create eth <> osmo pool, they can create one with no eth and osmo in their wallet as long as they have enough fee tokens?

File: CHANGELOG.md
Comment (p0mvn) [inline CHANGELOG.md:92]: I don't think we need a changelog entry for this. This is dev-focused, the end users should not be concerned about this change

```suggestion
```

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (p0mvn) [inline x/concentrated-liquidity/simulation/sim_msgs.go:183]: I think the reason why the failure is non-deterministic is because `math/rand` is used. We should use `rand := sim.GetRand()` to get a properly seeded rand AFAIU

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (p0mvn) [inline x/concentrated-liquidity/simulation/sim_msgs.go:6]: We should be using the randomness seeded in the simulator
```suggestion
```

File: x/concentrated-liquidity/simulation/sim_msgs.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/simulation/sim_msgs.go:338]: Correct

**Comment by czarcas7ic** [comment]
What was the decision on this? I think I remember something about us holding off on merging until something was figured out. Would like to note whatever the blocker is here so we can track it appropriatly.

**Comment by p0mvn** [comment]
> What was the decision on this? I think I remember something about us holding off on merging until something was figured out. Would like to note whatever the blocker is here so we can track it appropriately.

As far as I'm tracking, this is good to go as long as the non-deterministic failures are figured out and resolved.

If this is still failing from time to time, I think we should get to the bottom of this prior to merge

**Comment by czarcas7ic** [comment]
@stackman27 can you confirm if we are still having non-determinism in the test?

**Comment by stackman27** [comment]
> @stackman27 can you confirm if we are still having non-determinism in the test?

@p0mvn @czarcas7ic i was waiting on https://github.com/osmosis-labs/osmosis/pull/4973 to get merge. Will run some final test now and let you guys know

**Comment by stackman27** [comment]
> > @stackman27 can you confirm if we are still having non-determinism in the test?
> 
> @p0mvn @czarcas7ic i was waiting on #4973 to get merge. Will run some final test now and let you guys know

update on this 

i rebased and reran the tests. Everything seems good, haven't gotten the negative coin bug so far

**Review by p0mvn — COMMENTED** [review]
great work!

**Pull request merged**
